File: LeetCodePython/src/LengthOfLongestSubarrayWithAtMostKFrequency.py
# 2958
from typing import List


class Solution:

    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        ans = left = right = 0
        mydict = {}
        while left < nums.__len__():
            ans = max(ans, left - right)
            freq = mydict.get(nums[left], 0)
            if freq < k:
                mydict.update({nums[left]: freq + 1})
                left += 1
            else:
                right = self.ejectTillK(right, left, nums, mydict)
                mydict.update({nums[left]: freq - 1})
        ans = max(ans, left - right)
        return ans

    def ejectTillK(self, right: int, left: int, nums: List[int], mydict: dict) -> int:
        target = nums[left]
        flag = True
        while flag:
            curr = nums[right]
            currFreq = mydict.get(curr, 0)
            if currFreq == 1:
                mydict.pop(curr)
            else:
                mydict.update({curr: currFreq - 1})
            right += 1
            flag = target != curr
        return right

    print(maxSubarrayLength([1, 2, 3, 4, 5], 1))

File: LeetCodeGo/src/LengthOfLongestSubarrayWithAtMostKFrequency.go
package src

func maxSubarrayLength(nums []int, k int) int {
	left, right := 0, 0
	ans := 0
	myMap := make(map[int]int)
	for right < len(nums) {
		curr := nums[right]
		freq := myMap[curr]
		if freq < k {
			ans = max(ans, right-left)
			myMap[curr] = freq + 1
			right++
		} else {
			left = ejectTillK(left, right, nums, myMap)
		}
	}
	ans = max(ans, right-left)
	return ans
}

func ejectTillK(left int, right int, nums []int, myMap map[int]int) int {
	target := nums[right]
	flag := true
	for flag {
		curr := nums[left]
		currFreq := myMap[curr]
		if currFreq > 1 {
			myMap[curr] = currFreq - 1
		} else {
			delete(myMap, curr)
		}
		flag = curr != target
		left++
	}
	return left
}

func main() {
	input := []int{1, 2, 3, 4, 5}
	ans := maxSubarrayLength(input, 1)
	print(ans)
}
